### ISSUE TITLE
[FEA] Cache input data to reduce integration test time

### ISSUE THREAD

**Issue by ttnghia** [NVIDIA/spark-rapids#9763]
As of now, each running of integration tests can take more than 3 hours (more than 4 hours on databricks). We can consider caching the input data, storing all the randomly generated data in some static map and reusing them throughout the tests. By doing so, we could probably save a significant amount of time.

**Comment by ttnghia** [comment]
For example, if we have two tests using `IntegerGen()` then they can share exactly the same input.
Special generators such as `IntegerGen().with_special_case()` need to be considered as different from the pure `IntegerGen()` through internal object hashing.

**Comment by thirtiseven** [comment]
Currently we already have cached dataframes:
https://github.com/NVIDIA/spark-rapids/blob/30c3df35ab7e87f71ecd84789529d25d9a289848/integration_tests/src/main/python/data_gen.py#L758-L763

And have a `_cache_repr` in dataGens for hashing, is that the same thing as this issue?

**Comment by ttnghia** [comment]
Oh wow, I didn't know that. But we still have all tests running very slow. However, such decoration is only caching python lists of values, not Spark dataframes. Is it possible to cache Spark dataframes instead?

**Comment by thirtiseven** [comment]
> However, such decoration is only caching python lists of values, not Spark dataframes. Is it possible to cache Spark dataframes instead?

Yes, it is possible, we are not doing this now because the runtime config in spark session may affect the result dataframes. It is tracked by https://github.com/NVIDIA/spark-rapids/issues/8524
